🐛 Fix: Referral banner not shown on direct navigation to non-dashboard routes

niranjan-uma-shankar opened the pull request:
This PR fixes a bug that prevented the referral invite banner from appearing when users directly navigated to non-dashboard routes such as `/pages` or `/posts`. Since some users may have these routes bookmarked, it's important that the banner shows consistently across all eligible pages.

The issue is tracked in [#24448](https://github.com/TryGhost/Ghost/issues/24448).

### Root Cause
A typo in the role check ([isAdminOrOwnern](https://github.com/TryGhost/Ghost/blob/36b3e1c878f05273e558de36d29cef928227b79c/ghost/admin/app/components/gh-nav-menu/footer-banner.js#L56)) prevented the API call to fetch MRR data from being made on non-dashboard routes. Without MRR data, the banner's visibility condition was never satisfied.

### Fixes
* Corrects the typo in the conditional role check.
* Refactors loadCurrentMRR() to:
    - Include logic for applying mock data (from the Prototype Control Panel).
    - Ensure mock MRR data is loaded consistently across all routes, not just `/dashboard`.

This makes the banner behaviour more predictable and helps developers reliably test referral eligibility with mock data.



https://github.com/user-attachments/assets/dad4e08c-41af-4955-bc3b-1b4577397c7c


### History

The referral banner was first implemented in https://github.com/TryGhost/Ghost/pull/16187, and later copied over to the "What's new" banner implementation in https://github.com/TryGhost/Ghost/pull/20112. The typo exists in both implementations.


### Testing Instructions

1. Navigate to /dashboard and open the Prototype Control Panel.
    * Toggle on Enable fake data.
    * Set Emulated state to 90 days. This will ensure your site’s MRR exceeds $100.
2. Go to /settings/tiers and connect Stripe in test mode. At this point, you will become eligible to be shown the referral invite banner.
3. Go to `/dashboard` and confirm that the referral banner is visible in the nav menu.
    * The banner should remain visible as you navigate to other routes like `/pages` or `/posts`.
4. Now, directly open /pages in the browser (or any other non-dashboard route) and hard refresh the page using the browser reload button.
5. Verify that the referral invite banner loads in the nav menu correctly. 

<img width="1238" height="923" alt="Screenshot 2025-07-20 at 4 13 25 PM" src="https://github.com/user-attachments/assets/d867980d-931c-40d0-826a-838a656154f5" />

6. **Alternative way to verify the fix:**
    * First, check out the main branch and navigate to /pages.
    * Perform a hard reload and open your browser’s Network tab.
    * Observe that there is no API request made to /stats/mrr.

**Now**:

- Check out this branch (the one with the fix).
- Ensure that "Enable fake data" in the Prototype Control Panel is toggled off.
- Navigate again to /pages and hard reload.
-You should now see an outgoing request to `/stats/mrr` in the Network tab.

This confirms that MRR data is being fetched even when directly visiting non-dashboard routes, enabling the referral banner to appear for eligible accounts.

| Before | After |
|--------|-------|
| <img width="1012" height="947" alt="Before" src="https://github.com/user-attachments/assets/2f58c939-4464-4e7f-839e-45819c572349" /> | <img width="812" height="919" alt="After" src="https://github.com/user-attachments/assets/5cbd6c6d-668f-49d0-b00c-0f28bcf767f7" /> |


7. In your `ghost/admin/tests` folder, execute the following and verify that the tests pass:

```
yarn test 1 --filter="Referral Invite"
```

niranjan-uma-shankar committed:
Extracts logic to load saved mock state from localStorage

fixes https://github.com/TryGhost/Ghost/issues/24448
This file extracts the logic to load saved mock metadata, so that it can be reused in the footer banner component.

niranjan-uma-shankar committed:
Refactored loadState to use shared utility

fixes https://github.com/TryGhost/Ghost/issues/24448
To improve reusability and reduce duplication across modules.

niranjan-uma-shankar committed:
Returned default state shape when localStorage access fails

ref https://github.com/TryGhost/Ghost/issues/24448
Previously, the catch block returned an empty object, which could break consumers
expecting a consistent state shape. This change ensures a stable return value
even if localStorage is unavailable or contains malformed JSON.

niranjan-uma-shankar committed:
Moved mock state loading into loadCurrentMRR for banner visibility

ref https://github.com/TryGhost/Ghost/issues/24448
Included mock state loading in loadCurrentMRR to ensure MRR data is available,
which is required to determine eligibility for showing the referral invite banner.
This guarantees consistent initialization of MRR data and mock settings,
regardless of the initial route.

niranjan-uma-shankar commented on ghost/admin/app/components/gh-nav-menu/footer-banner.js:
```diff
@@ -53,7 +55,24 @@ export default class FooterBanner extends Component {
     @task
     *loadCurrentMRR() {
-        if (this.isAdminOrOwnern) {
+        if (this.isAdminOrOwner) {
```
Corrects typo

niranjan-uma-shankar committed:
Add line break

niranjan-uma-shankar committed:
Removed unused gh-referral-invite component files

ref https://github.com/TryGhost/Ghost/issues/24448
The files `gh-referral-invite.hbs` and `gh-referral-invite.js` under
`ghost/admin/app/components/` are no longer referenced anywhere in the codebase.
This component appears to be obsolete and can be safely removed to reduce
dead code and improve maintainability. The components were originally introduced in https://github.com/TryGhost/Ghost/pull/16187.

vershwal commented:
Thank you @niranjan-uma-shankar for this PR. Your change looks good. Could you please add some tests for this?

niranjan-uma-shankar committed:
Acceptance tests added for the Referral Invite banner

To ensure that the referral invite banner behaves correctly when an account is eligible to be shown the banner.

niranjan-uma-shankar committed:
EOL

niranjan-uma-shankar committed:
Undo previous commit

niranjan-uma-shankar committed:
Refactor for readability

Uses object property shorthand for improved readability.

niranjan-uma-shankar committed:
Removes unused moment import

Moment is not used in the file.
